### PR TITLE
Allow drag and drop of songs to favorite playlists

### PR DESCRIPTION
**Pull request by bespsm**
outstrawberry

- allows adding songs from active playlist to any favourite by drag&drop
- after 500msec hovering with the songs over desired playlist it becomes current
- drag & drop multiple songs is supported

### CODE REVIEW

File: src/playlist/playlistlistcontainer.cpp [hunk 1]
```diff
@@ -348,6 +350,20 @@ void PlaylistListContainer::ItemDoubleClicked(const QModelIndex &proxy_idx) {
 
 }
 
+void PlaylistListContainer::ItemMimeDataDropped(const QModelIndex &proxy_idx, const QMimeData *q_mimedata) {
+
+  const QModelIndex idx = proxy_->mapToSource(proxy_idx);
+  if (!idx.isValid()) return;
+
+  // Drop playlist rows if type is playlist and
```
Comment (jonaski) [inline src/playlist/playlistlistcontainer.cpp:358]: Comment on same line

File: src/playlist/playlistlistcontainer.cpp [hunk 2]
```diff
@@ -348,6 +350,20 @@ void PlaylistListContainer::ItemDoubleClicked(const QModelIndex &proxy_idx) {
...
+  // it's not active, to prevent selfcopy
+  int playlistId = idx.data(PlaylistListModel::Role_PlaylistId).toInt();
+  if (idx.data(PlaylistListModel::Role_Type).toInt() == PlaylistListModel::Type_Playlist
```
Comment (jonaski) [inline src/playlist/playlistlistcontainer.cpp:361]: one line

File: src/playlist/playlistlistview.h [hunk 3]
```diff
@@ -24,29 +24,47 @@
...
+class QDropEvent;
+class QTimerEvent;
 
 class PlaylistListView : public AutoExpandingTreeView {
   Q_OBJECT
 
  public:
   explicit PlaylistListView(QWidget *parent = nullptr);
 
+  static const int kDragHoverTimeout = 500;
```
Comment (jonaski) [inline src/playlist/playlistlistview.h:47]: This should be private

File: src/playlist/playlistlistview.h [hunk 4]
```diff
@@ -24,29 +24,47 @@
...
+
   bool ItemsSelected() const;
 
  signals:
   void ItemsSelectedChanged(const bool);
+  void ItemMimeDataDroppedSignal(const QModelIndex &proxy_idx, const QMimeData *q_mimedata);
```
Comment (jonaski) [inline src/playlist/playlistlistview.h:53]: No const for q_mimedata

File: src/playlist/playlistlistview.h [hunk 5]
```diff
@@ -24,29 +24,47 @@
...
   void paintEvent(QPaintEvent *event) override;
   void selectionChanged(const QItemSelection&, const QItemSelection&) override;
+
+  void dragEnterEvent(QDragEnterEvent *e) override;
+  void dragMoveEvent(QDragMoveEvent *e) override;
+  void dragLeaveEvent(QDragLeaveEvent *e) override;
+  void dropEvent(QDropEvent *e) override;
+  void timerEvent(QTimerEvent *e) override;
+
+private:
```
Comment (jonaski) [inline src/playlist/playlistlistview.h:66]:
```suggestion
 private:
```

File: src/playlist/playlistlistview.cpp [hunk 6]
```diff
@@ -64,3 +66,65 @@ bool PlaylistListView::ItemsSelected() const {
...
+  if (e->mimeData()->hasFormat(Playlist::kRowsMimetype)) {
+    if (drag_hover_timer_.isActive()) {
+      drag_hover_timer_.stop();
+    }
+    emit ItemMimeDataDroppedSignal(currentIndex(), e->mimeData());
+  }
+  else  {
+    AutoExpandingTreeView::dropEvent(e);
+  }
+}
```
Comment (jonaski) [inline src/playlist/playlistlistview.cpp:130]: Newline before }

File: src/playlist/playlistlistview.cpp [hunk 7]
```diff
@@ -64,3 +66,65 @@ bool PlaylistListView::ItemsSelected() const {
...
 }
+
+void PlaylistListView::dragEnterEvent(QDragEnterEvent *e) {
+  if (e->mimeData()->hasFormat(Playlist::kRowsMimetype)) {
+    e->acceptProposedAction();
+  }
+  else {
+    AutoExpandingTreeView::dragEnterEvent(e);
+  }
+}
```
Comment (jonaski) [inline src/playlist/playlistlistview.cpp:77]: Newline before }

File: src/playlist/playlistlistview.cpp [hunk 8]
```diff
@@ -64,3 +66,65 @@ bool PlaylistListView::ItemsSelected() const {
...
+      if (drag_hover_timer_.isActive()) {
+        drag_hover_timer_.stop();
+      }
+      drag_hover_timer_.start(kDragHoverTimeout, this);
+    }
+  }
+  else {
+    AutoExpandingTreeView::dragMoveEvent(e);
+  }
+}
```
Comment (jonaski) [inline src/playlist/playlistlistview.cpp:101]: Newline before }

File: src/playlist/playlistlistview.cpp [hunk 9]
```diff
@@ -64,3 +66,65 @@ bool PlaylistListView::ItemsSelected() const {
...
+
+void PlaylistListView::timerEvent(QTimerEvent *e) {
+
+  QTreeView::timerEvent(e);
+  if (e->timerId() == drag_hover_timer_.timerId()) {
+    drag_hover_timer_.stop();
+    emit doubleClicked(currentIndex());
+  }
+}
```
Comment (jonaski) [inline src/playlist/playlistlistview.cpp:117]: Newline before }

File: src/playlist/playlistlistcontainer.cpp [hunk 10]
```diff
@@ -348,6 +350,20 @@ void PlaylistListContainer::ItemDoubleClicked(const QModelIndex &proxy_idx) {
 
 }
 
+void PlaylistListContainer::ItemMimeDataDropped(const QModelIndex &proxy_idx, const QMimeData *q_mimedata) {
```
Comment (jonaski) [inline src/playlist/playlistlistcontainer.cpp:353]: No const for q_mimedata

File: src/playlist/playlistlistcontainer.cpp [hunk 11]
```diff
@@ -348,6 +350,20 @@ void PlaylistListContainer::ItemDoubleClicked(const QModelIndex &proxy_idx) {
...
+  if (!idx.isValid()) return;
+
+  // Drop playlist rows if type is playlist and
+  // it's not active, to prevent selfcopy
+  int playlistId = idx.data(PlaylistListModel::Role_PlaylistId).toInt();
+  if (idx.data(PlaylistListModel::Role_Type).toInt() == PlaylistListModel::Type_Playlist
+    && playlistId != app_->playlist_manager()->active_id()) {
+    app_->playlist_manager()->playlist(playlistId)->dropMimeData(q_mimedata, Qt::CopyAction, -1, 0, QModelIndex());
+  }
+}
```
Comment (jonaski) [inline src/playlist/playlistlistcontainer.cpp:365]: Newline before }

File: src/playlist/playlistlistview.cpp [hunk 12]
```diff
@@ -64,3 +66,65 @@ bool PlaylistListView::ItemsSelected() const {
 void PlaylistListView::selectionChanged(const QItemSelection&, const QItemSelection&) {
   emit ItemsSelectedChanged(selectionModel()->selectedRows().count() > 0);
 }
+
+void PlaylistListView::dragEnterEvent(QDragEnterEvent *e) {
```
Comment (jonaski) [inline src/playlist/playlistlistview.cpp:70]: Newline after {

File: src/playlist/playlistlistcontainer.cpp [hunk 13]
```diff
@@ -348,6 +350,20 @@ void PlaylistListContainer::ItemDoubleClicked(const QModelIndex &proxy_idx) {
...
 }
 
+void PlaylistListContainer::ItemMimeDataDropped(const QModelIndex &proxy_idx, const QMimeData *q_mimedata) {
+
+  const QModelIndex idx = proxy_->mapToSource(proxy_idx);
+  if (!idx.isValid()) return;
+
+  // Drop playlist rows if type is playlist and
+  // it's not active, to prevent selfcopy
+  int playlistId = idx.data(PlaylistListModel::Role_PlaylistId).toInt();
```
Comment (jonaski) [inline src/playlist/playlistlistcontainer.cpp:360]: playlist_id

File: src/playlist/playlistlistview.cpp [hunk 14]
```diff
@@ -64,3 +66,65 @@ bool PlaylistListView::ItemsSelected() const {
...
+      }
+      drag_hover_timer_.start(kDragHoverTimeout, this);
+    }
+  }
+  else {
+    AutoExpandingTreeView::dragMoveEvent(e);
+  }
+}
+
+void PlaylistListView::dragLeaveEvent(QDragLeaveEvent *e) {
```
Comment (jonaski) [inline src/playlist/playlistlistview.cpp:103]: Newline after {

File: src/playlist/playlistlistview.cpp [hunk 15]
```diff
@@ -64,3 +66,65 @@ bool PlaylistListView::ItemsSelected() const {
...
+  if (drag_hover_timer_.isActive()) {
+    drag_hover_timer_.stop();
+  }
+  AutoExpandingTreeView::dragLeaveEvent(e);
+}
```
Comment (jonaski) [inline src/playlist/playlistlistview.cpp:108]: Newline before }

**Review by jonaski — CHANGES_REQUESTED** [review]
Thanks. A few minor code-formatting adjustments.

Please also update commit message, no "feature:" prefix and fix typo. suggestion:
"Allow drag and drop of songs to favorite playlists".

**Comment by jonaski** [comment]
Could you rebase this with master, I fixed an error in the CI yesterday related to the macOS build failing because it does not have access to the GH secrets from forks. (https://github.com/strawberrymusicplayer/strawberry/commit/c52fc90306990025bbf0477ffad42e87d268d88d)

File: src/playlist/playlistlistview.h (hunk 4 above)
Comment (bespsm) [reply src/playlist/playlistlistview.h:51]: are you sure? `const` came from this method https://doc.qt.io/qt-6/qdropevent.html#mimeData

I have to cast it...

Comment (jonaski) [reply src/playlist/playlistlistview.h:51]: OK, keep const then.

File: src/playlist/playlistlistview.cpp (hunk 7 above)
Comment (bespsm) [reply src/playlist/playlistlistview.cpp:79]: added

File: src/playlist/playlistlistview.cpp (hunk 8 above)
Comment (bespsm) [reply src/playlist/playlistlistview.cpp:104]: added

File: src/playlist/playlistlistview.cpp (hunk 9 above)
Comment (bespsm) [reply src/playlist/playlistlistview.cpp:123]: added

File: src/playlist/playlistlistview.cpp (hunk 6 above)
Comment (bespsm) [reply src/playlist/playlistlistview.cpp:137]: added

File: src/playlist/playlistlistcontainer.cpp (hunk 10 above)
Comment (bespsm) [reply src/playlist/playlistlistcontainer.cpp:353]: already discussed

**Review by jonaski — APPROVED** [review]

**Pull request merged**